File: hacks/pretty-c++/src/film_view.cc
FilmView::FilmView ()
	: _film (0)
	, _content_file_radio (_content_group)
	, _content_file_chooser ("Content", Gtk::FILE_CHOOSER_ACTION_OPEN)
	, _content_file_button (_content_file_chooser)
	, _content_folder_radio (_content_group)
	, _content_folder_chooser ("Content", Gtk::FILE_CHOOSER_ACTION_SELECT_FOLDER)
	, _content_folder_button (_content_folder_chooser)
	, _dvd ("DVD")
	, _deinterlace ("Deinterlace")
{
	_table = manage (new Gtk::Table);
	table->set_row_spacings (4);
	table->set_col_spacings (12);
	main_vbox->pack_start (*table, false, false);

	int n = 0;

	table->attach (*left_aligned_label ("Content"), 0, 1, n, n + 1);
	_content_file_chooser.add_button ("Select", Gtk::RESPONSE_OK);
	_content_folder_chooser.add_button ("Select", Gtk::RESPONSE_OK);
	Gtk::HBox* b = Gtk::manage (new Gtk::HBox);
	b->set_spacing (12);
	_content_file_radio.set_label ("File");
	b->pack_start (_content_file_radio, false, false);
	b->pack_start (_content_file_button, true, true);
	_content_folder_radio.set_label ("Folder");
	b->pack_start (_content_folder_radio, false, false);
	b->pack_start (_content_folder_button, true, true);
	table->attach (*b, 1, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Name"), 0, 1, n, n + 1);
	table->attach (_name, 1, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Ratio"), 0, 1, n, n + 1);
	_ratio.append_text ("1.33:1 (4:3)");
	_ratio.append_text ("1.78:1 (16:9)");
	_ratio.append_text ("1.85:1 (Flat)");
	_ratio.append_text ("2.39:1 (Scope)");
	table->attach (_ratio, 1, 2, n, n + 1);
	_ratio.set_active_text ("1.85:1 (Flat)");
	++n;

	table->attach (_dvd, 0, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("DVD title"), 0, 1, n, n + 1);
	_dvd_title.set_range (1, 64);
	_dvd_title.set_increments (1, 4);
	_dvd_title.set_value (1);
	table->attach (_dvd_title, 1, 2, n, n + 1);
	++n;

	table->attach (_deinterlace, 0, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Left Crop"), 0, 1, n, n + 1);
	_left_crop.set_range (0, 1024);
	_left_crop.set_increments (1, 64);
	_left_crop.set_value (0);
	table->attach (_left_crop, 1, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Right Crop"), 0, 1, n, n + 1);
	_right_crop.set_range (0, 1024);
	_right_crop.set_increments (1, 64);
	_right_crop.set_value (0);
	table->attach (_right_crop, 1, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Top Crop"), 0, 1, n, n + 1);
	_top_crop.set_range (0, 1024);
	_top_crop.set_increments (1, 64);
	_top_crop.set_value (0);
	table->attach (_top_crop, 1, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Bottom Crop"), 0, 1, n, n + 1);
	_bottom_crop.set_range (0, 1024);
	_bottom_crop.set_increments (1, 64);
	_bottom_crop.set_value (0);
	table->attach (_bottom_crop, 1, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Size"), 0, 1, n, n + 1);
	table->attach (_size, 1, 2, n, n + 1);
	++n;

	table->attach (*left_aligned_label ("Length"), 0, 1, n, n + 1);
	table->attach (_length, 1, 2, n, n + 1);
	++n;
	
	/* We only need to connect to one of the radios in the group */
	_content_file_radio.signal_toggled().connect (sigc::mem_fun (*this, &FilmView::content_radio_toggled));
	_content_file_button.signal_file_set().connect (sigc::mem_fun (*this, &FilmView::content_changed));
	_content_folder_button.signal_file_set().connect (sigc::mem_fun (*this, &FilmView::content_changed));
	_dvd.signal_toggled().connect (sigc::mem_fun (*this, &FilmView::update_dvd_title_sensitivity));

	update_content_radio_sensitivity ();
	update_dvd_title_sensitivity ();
	show_all ();
}

Gtk::Label *
FilmView::left_aligned_label (string const & t) const
{
	Gtk::Label* l = Gtk::manage (new Gtk::Label (t));
	l->set_alignment (0, 0.5);
	return l;
}

void
FilmView::content_radio_toggled ()
{
	update_content_radio_sensitivity ();
	content_changed ();
}

void
FilmView::update_content_radio_sensitivity ()
{
	_content_file_button.set_sensitive (_content_file_radio.get_active ());
	_content_folder_button.set_sensitive (_content_folder_radio.get_active ());
}

void
FilmView::update_dvd_title_sensitivity ()
{
	_dvd_title.set_sensitive (_dvd.get_active ());
}

void
FilmView::content_changed ()
{
	cout << "Content changed.\n";
}

void
FilmView::set_film (Film* f)
{
	_film = f;
	update ();
}

File: hacks/python-playback/film_view.py
import os
import pygtk
pygtk.require('2.0')
import gtk
import ratio
import util
import thumbs

class FilmView(gtk.HBox):
    def __init__(self, parent):
        gtk.HBox.__init__(self)

        self.parent_window = parent

        self.inhibit_save = True
        
        self.table = gtk.Table()
        self.pack_start(self.table, True, True)

        self.table.set_row_spacings(4)
        self.table.set_col_spacings(12)
        self.film_name = gtk.Entry()
        self.ratio = gtk.combo_box_new_text()
        for r in ratio.ratios:
            self.ratio.append_text(r.name())
        self.content_file_radio = gtk.RadioButton()
        self.content_file_radio.set_label("File")
        self.content_file_chooser = gtk.FileChooserDialog("Content", self.parent_window, gtk.FILE_CHOOSER_ACTION_OPEN, (("Select", gtk.RESPONSE_OK)))
        self.content_file_button = gtk.FileChooserButton(self.content_file_chooser)
        self.content_folder_radio = gtk.RadioButton()
        self.content_folder_radio.set_label("Folder")
        self.content_folder_radio.set_group(self.content_file_radio)
        self.content_folder_chooser = gtk.FileChooserDialog("Content", self.parent_window, gtk.FILE_CHOOSER_ACTION_SELECT_FOLDER, (("Select", gtk.RESPONSE_OK)))
        self.content_folder_button = gtk.FileChooserButton(self.content_folder_chooser)
        self.dvd = gtk.CheckButton("DVD")
        self.dvd_title = gtk.SpinButton()
        self.dvd_title.set_range(0, 32)
        self.dvd_title.set_increments(1, 4)
        self.deinterlace = gtk.CheckButton("Deinterlace")
        self.left_crop = self.crop_spinbutton()
        self.right_crop = self.crop_spinbutton()
        self.top_crop = self.crop_spinbutton()
        self.bottom_crop = self.crop_spinbutton()
        
        # Information about the content (immutable)
        self.source_size = self.label()
        self.fps = self.label()
        self.length = self.label()

        # Buttons
        self.thumbs_button = gtk.Button("Show Thumbnails")

        self.film_name.connect("changed", self.changed, self)
        self.ratio.connect("changed", self.changed, self)
        self.deinterlace.connect("toggled", self.changed, self)
        self.thumbs_button.connect("clicked", self.thumbs_clicked, self)
        self.content_file_radio.connect("toggled", self.content_radio_toggled, self)
        self.content_folder_radio.connect("toggled", self.content_radio_toggled, self)
        self.content_file_button.connect("file-set", self.content_file_chooser_file_set, self)
        self.content_folder_button.connect("file-set", self.content_folder_chooser_file_set, self)
        self.dvd.connect("toggled", self.changed, self)
        self.dvd_title.connect("value-changed", self.changed, self)

        n = 0
        self.table.attach(self.label("Film"), 0, 1, n, n + 1)
        self.table.attach(self.film_name, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Ratio"), 0, 1, n, n + 1)
        self.table.attach(self.ratio, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Content"), 0, 1, n, n + 1)
        b = gtk.HBox()
        b.set_spacing(4)
        b.pack_start(self.content_file_radio, False, False)
        b.pack_start(self.content_file_button, True, True)
        b.pack_start(self.content_folder_radio, False, False)
        b.pack_start(self.content_folder_button, True, True)
        self.table.attach(b, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.dvd, 0, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("DVD title"), 0, 1, n, n + 1)
        self.table.attach(self.dvd_title, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.deinterlace, 0, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Left Crop"), 0, 1, n, n + 1)
        self.table.attach(self.left_crop, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Right Crop"), 0, 1, n, n + 1)
        self.table.attach(self.right_crop, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Top Crop"), 0, 1, n, n + 1)
        self.table.attach(self.top_crop, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Bottom Crop"), 0, 1, n, n + 1)
        self.table.attach(self.bottom_crop, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Source size"), 0, 1, n, n + 1)
        self.table.attach(self.source_size, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Frames per second"), 0, 1, n, n + 1)
        self.table.attach(self.fps, 1, 2, n, n + 1)
        n += 1
        self.table.attach(self.label("Length"), 0, 1, n, n + 1)
        self.table.attach(self.length, 1, 2, n, n + 1)

        self.right_vbox = gtk.VBox()
        self.pack_start(self.right_vbox, False, False)

        self.right_vbox.pack_start(self.thumbs_button, False, False)

        self.inhibit_save = False

    def set(self, film):
        self.inhibit_save = True

        self.film = film
        self.film_name.set_text(film.name)
        self.ratio.set_active(ratio.ratio_to_index(film.ratio))
        if film.content is not None:
            if os.path.isfile(film.content):
                self.set_content_is_file(True)
                self.content_file_button.set_filename(film.content)
            else:
                self.set_content_is_file(False)
                self.content_folder_button.set_filename(film.content)
        self.dvd.set_active(film.dvd)
        self.dvd_title.set_value(film.dvd_title)
        self.dvd_title.set_sensitive(film.dvd)
        self.deinterlace.set_active(film.deinterlace)
        self.left_crop.set_value(film.left_crop)
        self.right_crop.set_value(film.right_crop)
        self.top_crop.set_value(film.top_crop)
        self.bottom_crop.set_value(film.bottom_crop)

        # Content information
        if film.width is not None and film.height is not None:
            self.source_size.set_text("%dx%d" % (film.width, film.height))
        else:
            self.source_size.set_text("Unknown")
        if film.fps is not None:
            self.fps.set_text(str(film.fps))
        if film.length is not None:
            self.length.set_text("%d:%02d:%02d" % util.s_to_hms(film.length))

        self.inhibit_save = False

    def set_content_is_file(self, f):
        self.content_file_button.set_sensitive(f)
        self.content_folder_button.set_sensitive(not f)
        self.content_file_radio.set_active(f)
        self.content_folder_radio.set_active(not f)
    
    def label(self, text = ""):
        l = gtk.Label(text)
        l.set_alignment(0, 0.5)
        return l

    def changed(self, a, b):
        self.dvd_title.set_sensitive(self.dvd.get_active())
        self.save_film()

    def crop_spinbutton(self):
        s = gtk.SpinButton()
        s.set_range(0, 1024)
        s.set_increments(1, 16)
        s.connect("value-changed", self.changed, self)
        return s

    def save_film(self):
        if self.inhibit_save:
            return

        self.film.name = self.film_name.get_text()
        self.film.ratio = ratio.index_to_ratio(self.ratio.get_active()).ratio

        if self.content_file_radio.get_active():
            self.film.set_content(self.content_file_button.get_filename())
        else:
            self.film.set_content(self.content_folder_button.get_filename())
        self.film.set_dvd(self.dvd.get_active())
        self.film.set_dvd_title(self.dvd_title.get_value_as_int())
        self.film.deinterlace = self.deinterlace.get_active()
        self.film.left_crop = self.left_crop.get_value_as_int()
        self.film.right_crop = self.right_crop.get_value_as_int()
        self.film.top_crop = self.top_crop.get_value_as_int()
        self.film.bottom_crop = self.bottom_crop.get_value_as_int()
        self.film.write()

    def thumbs_clicked(self, a, b):
        if self.film.thumbs() == 0:
            self.film.make_thumbs()

        t = thumbs.Thumbs(self.film)
        t.run()
        t.hide()
        self.left_crop.set_value(t.left_crop())
        self.right_crop.set_value(t.right_crop())
        self.top_crop.set_value(t.top_crop())
        self.bottom_crop.set_value(t.bottom_crop())

    def content_file_chooser_file_set(self, a, b):
        self.changed(a, b)

    def content_folder_chooser_file_set(self, a, b):
        self.changed(a, b)

    def content_radio_toggled(self, a, b):
        self.set_content_is_file(self.content_file_radio.get_active())
        self.changed(a, b)

